src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp

```diff
@@ -0,0 +1,350 @@
+    {
+        throw ngraph_error(
+            "Max-pool window shape rank does not match number of spatial dimensions.");
+    }
+
+    if (window_movement_strides.size() != spatial_dimension_count)
+    {
+        throw ngraph_error(
+            "Max-pool window movement stride rank does not match number of spatial "
+            "dimensions.");
+    }
+
+    if (padding_below.size() != spatial_dimension_count)
+    {
+        throw ngraph_error(
+            "Max-pool below-padding rank does not match number of spatial dimensions.");
+    }
+
+    if (padding_above.size() != spatial_dimension_count)
+    {
+        throw ngraph_error(
+            "Max-pool above-padding rank does not match number of spatial dimensions.");
+    }
+
+    //
+    // Extract input item shape Di and make sure all dimensions are larger than 0.
+    //
+    Shape input_item_virtual_shape;
+
+    for (size_t i = 0; i < spatial_dimension_count; i++)
+    {
+        size_t dim_size = arg_shape[1 + 1 + i];
+        size_t virtual_dim_size = padding_below[i] + dim_size + padding_above[i];
+        input_item_virtual_shape.push_back(virtual_dim_size);
+
+        if (virtual_dim_size == 0)
+        {
+            throw ngraph_error("Max-pool input spatial dimension is zero even after padding.");
+        }
+    }
+
+    //
+    // Make sure window shape dimensions are all larger than 0.
+    //
+    for (size_t i = 0; i < spatial_dimension_count; i++)
+    {
+        if (window_shape[i] == 0)
+        {
+            throw ngraph_error("Max-pool window shape has a zero-length axis.");
+        }
+    }
+
+    //
+    // Make the max pooling window fits within the spatial dimensions.
+    //
+    for (size_t i = 0; i < spatial_dimension_count; i++)
+    {
+        if (window_shape[i] > input_item_virtual_shape[i])
+        {
+            throw ngraph_error(
+                "Max-pool window shape is larger than the spatial dimensions even after "
+                "padding.");
+        }
+    }
+
+    //
+    // Compute output item shape Do, checking at the same time that all window movement strides are larger than 0.
+    //
+    Shape output_item_shape;
+
+    for (size_t i = 0; i < spatial_dimension_count; i++)
+    {
+        if (window_movement_strides[i] == 0)
+        {
+            throw ngraph_error("Max-pool window axis movement stride is zero.");
+        }
+        output_item_shape.push_back(ceil_div(input_item_virtual_shape[i] - window_shape[i] + 1,
+                                             window_movement_strides[i]));
+    }
+
+    //
+    // Construct result shape: NCDo.
+    //
+    Shape result_shape(1 + 1 + spatial_dimension_count);
+    result_shape[0] = batch_size;
+    result_shape[1] = channel_count;
+    copy(output_item_shape.begin(), output_item_shape.end(), result_shape.begin() + 2);
+
+    add_output(get_input_element_type(0), result_shape);
+    add_output(element::i32, result_shape);
```

Comment:
We will need a defintion/comment about the value of indices. Get the definition from mkldnn folks. Depending on that we can also define the element type.

Reply:
MKLDNN uses either `u8` or `s32` to encode indices , [source](https://github.com/intel/mkl-dnn/blob/2df65197f1c841473144a6fa7e3118628baac773/src/cpu/cpu_pooling_pd.hpp#L35) I'd suggest we could always use `i32` until there's some evidence that we need to optimize for space or to reduce copying.

Reply:
i meant this: https://github.com/intel/mkl-dnn/blob/2df65197f1c841473144a6fa7e3118628baac773/src/cpu/ref_pooling.cpp#L92 
'kh*KW + kw'

TF MaxPoolwithArgMax has a different definition of ArgMax.